File: FlinkScalaCode/src/main/scala/com/wubaibao/flinkscala/code/chapter6/sink/RedisSinkTest.scala
package com.wubaibao.flinkscala.code.chapter6.sink

import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.streaming.connectors.redis.RedisSink
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig
import org.apache.flink.streaming.connectors.redis.common.mapper.{RedisCommand, RedisCommandDescription, RedisMapper}

/**
 * Flink Redis Sink 测试
 */
object RedisSinkTest {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    //设置隐式转换
    import org.apache.flink.api.scala._

    /**
     * Socket 中输入数据格式：
     * hello,world
     * hello,flink
     * hello,scala
     * hello,spark
     * hello,hadoop
     */
    val ds: DataStream[String] = env.socketTextStream("node5", 9999)

    //统计wordCount
    val result: DataStream[(String, Int)] = ds.flatMap(_.split(","))
      .map((_, 1))
      .keyBy(_._1)
      .sum(1)

    //准备RedisSink对象
    val config: FlinkJedisPoolConfig = new FlinkJedisPoolConfig.Builder()
      .setHost("node4")
      .setPort(6379)
      .setDatabase(1)
      .build()
    val redisSink = new RedisSink[(String, Int)](config, new RedisMapper[(String, Int)] {
      //指定写入Redis的命令
      override def getCommandDescription: RedisCommandDescription = {
        new RedisCommandDescription(RedisCommand.HSET, "flink-scala-redis")
      }

      //指定写入Redis的Key
      override def getKeyFromData(t: (String, Int)): String = t._1

      //指定写入Redis的Value
      override def getValueFromData(t: (String, Int)): String = t._2.toString
    })

    //将结果写入Redis
    result.addSink(redisSink)
    env.execute("RedisSinkTest")



  }

}

File: FlinkJavaCode/src/main/java/com/wubaibao/flinkjava/code/chapter6/sink/RedisSinkTest.java
package com.wubaibao.flinkjava.code.chapter6.sink;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.redis.RedisSink;
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommand;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommandDescription;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisMapper;

/**
 * Flink Redis Sink 测试
 */
public class RedisSinkTest {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        /**
         * Socket中输入数据如下：
         * hello,flink
         * hello,spark
         * hello,hadoop
         * hello,java
         */
        DataStreamSource<String> ds1 = env.socketTextStream("node5", 9999);

        //统计wordcount
        SingleOutputStreamOperator<Tuple2<String, Integer>> result =
                ds1.flatMap((FlatMapFunction<String, String>) (s, collector) -> {
                    String[] arr = s.split(",");
                    for (String word : arr) {
                        collector.collect(word);
                    }
                }).returns(Types.STRING)
                .map(one -> Tuple2.of(one, 1)).returns(Types.TUPLE(Types.STRING, Types.INT))
                .keyBy(tp -> tp.f0)
                .sum(1);

        //准备RedisSink对象
        FlinkJedisPoolConfig conf = new FlinkJedisPoolConfig.Builder()
                .setHost("node4")
                .setPort(6379)
                .setDatabase(1)
                .build();

        RedisSink<Tuple2<String, Integer>> redisSink = new RedisSink<>(conf, new RedisMapper<Tuple2<String, Integer>>() {
            @Override
            public RedisCommandDescription getCommandDescription() {
                //指定Redis命令描述，不需预先创建Redis表
                return new RedisCommandDescription(RedisCommand.HSET, "flink-redis-sink");
            }

            @Override
            public String getKeyFromData(Tuple2<String, Integer> tp) {
                //指定Redis Key
                return tp.f0;
            }

            @Override
            public String getValueFromData(Tuple2<String, Integer> tp) {
                //指定Redis Value
                return tp.f1 + "";
            }
        });

        //将结果写入Redis
        result.addSink(redisSink);
        env.execute();


    }
}
